src/foam/dao/MaterializedDAO.js

```diff
@@ -250,54 +260,83 @@ foam.CLASS({
         return this;
       `
     },
-
     {
       name: 'run',
       type: 'void',
       javaCode: `
-        Object[] cmd;
-        FObject  value;
-
         foam.core.XLocator.set(getX());
-
-        // TODO: Support two modes: SERIAL and PARALLEL
-        // SERIAL mode is as below
-        // PARALLEL mode only has to handle PUT, can just submit
-        // to threadpool
-        // Transition from PARALLEL to SERIAL needs to block until
-        // pool is drained. Might be easier to have own pool to make
-        // this easier. Or maintain a count.
         while ( true ) {
           try {
-            cmd = (Object[]) getQueue().take();
-
-            if ( cmd[0] == PUT ) {
-              value = (FObject) cmd[1];
+            process((Object[]) getQueue().take());
+          } catch (InterruptedException e) {}
+        }
+      `
+    },
+    {
+      name: 'process',
+      args: 'Object[] cmd',
+      javaCode: `
+        FObject  value;
+        if ( cmd[0] == PUT ) {
+          value = (FObject) cmd[1];
 
-              if ( getPredicate().f(value) ) {
-                var obj = adapt(value);
-                if ( obj != null )
-                  getDelegate().put(obj);
-              }
-            } else if ( cmd[0] == REMOVE ) {
-              value = (FObject) cmd[1];
+          if ( getPredicate().f(value) ) {
+            var obj = adapt(value);
+            if ( obj != null )
+              getDelegate().put(obj);
+          }
+        } else if ( cmd[0] == REMOVE ) {
+          value = (FObject) cmd[1];
 
-              if ( getPredicate().f(value) ) {
-                var obj = getAdapter().fastAdapt(value);
-                if ( obj != null )
-                  getDelegate().remove(obj);
-              }
-            } else /* removeAll */ {
-              getDelegate().removeAll();
-            }
-          } catch (InterruptedException e) {}
+          if ( getPredicate().f(value) ) {
+            var obj = getAdapter().fastAdapt(value);
+            if ( obj != null )
+              getDelegate().remove(obj);
+          }
+        } else /* removeAll */ {
+          getDelegate().removeAll();
         }
       `
     },
     {
       name: 'start',
       javaCode: `
-      if ( getAutoStart() ) maybeInit();
+      if ( getAutoStart() ) {
+        synchronized ( this ) {
+          if ( getInitialized() ||
+               getInitializing() ) return;
+          setInitializing(true);
+       }
+       Logger logger = Loggers.logger(getX(), this, getSourceDAO().getOf().getObjClass().getSimpleName(), "start");
+
+       foam.core.XLocator.set(getX());
+       try {
+          long count = ((Count) getSourceDAO().select(new Count())).getValue();
```

Comment:
I'm aware of this, but don't believe it's an issue.  The puts are queued.  If count is low the run thread will take care of it.  
The only scenario I can think of is a delete after the count.
Perhaps just a lock.  The index commands generally only occur on first use.  So there would be no lock use during normal operation.

Reply:
The addIndex() is already under lock, which is why my idea was to do everything as one operation under that lock to avoid having to do anything else. Where the Index is fed the data already knows the size and is under lock so it would just need to tell the index when it's done the same-to-handle-concurrently part.

Reply:
This would let any Index potentially optimize when it knew there wouldn't be any updates or deletes.

Reply:
A simple hackish fix is to just remove something like 10 or 100 from the count.